pkg/wal/wal.go

```diff
@@ -0,0 +1,345 @@
+	<- w.connectionControl
+	return
+}
+
+type walChannels struct {
+	tokens  chan []string
+	entry   chan *Entry
+	entries chan []Entry
+	count   chan int
+	oops    chan error
+	done chan struct{}
+}
+
+func newWalChannels() *walChannels {
+	token := make(chan []string)
+	entry := make(chan *Entry)
+	entries := make(chan []Entry)
+	count := make (chan int)
+	done := make(chan struct{})
+	return &walChannels{
+		tokens:  token,
+		entry:   entry,
+		entries: entries,
+		count:   count,
+		done:    done,
+	}
+}
+
+func (w *WAL) ListEntries(ctx context.Context, fromToken string, max int) ([]Entry, string,  error) {
+
+	walChannels := newWalChannels()
+	count := 0
+
+	// Start routine which collects all parallel responses.
+	go w.collectParallelResponses(ctx, walChannels)
+	// Start routine that will issue parallel reads.
+	go w.issueParallelReads(ctx, walChannels)
+
+	// Read tokens and send to routine to entry in parallel while fetching new tokens.
+	for count = 0; count < max; {
+		var tokens []string
+
+		tokens, next, err := w.ListTokens(ctx, fromToken, max-count)
+
+		if err != nil {
+			return nil, "", fmt.Errorf("wal failed to ListEntries: " + err.Error())
+		}
+
+		count += len(tokens)
+
+		if len(tokens) > 0 {
+			// account for case list is not complete but the response does not include any token.
+			walChannels.tokens <- tokens
+		}
+
+		fromToken = next
+
+		// Read max or all
+		if (count == max) || (fromToken == "") {
+			break
+		}
+	}
+
+	close(walChannels.tokens) // No more tokens to report.
+
+	walChannels.count <- count // Send total count for reading responses.
+
+	for {
+		select {
+		case entries := <- walChannels.entries:
+			if len(entries) != count {
+				panic(fmt.Sprintf("received count different than the list of tokens entries: %d count: %d", len(entries), count))
+			}
+			return entries, fromToken, nil
+		case err := <- walChannels.oops:
+			return nil, "", err
+		}
+	}
+}
+
+func (w *WAL) issueParallelReads(ctx context.Context, channels *walChannels)  {
+	count := 0
+	for {
+		select {
+		case tokens, ok := <- channels.tokens:
+			if !ok {
+				// Closed and empty: https://golang.org/ref/spec#Receive_operator
+				w.l.Info("Completed issuing parallel reads", zap.Int("count", count))
+				return
+			}
+			for _, t := range tokens {
+				w.getConnection() // concurrency control
+				count++
+				go w.read(ctx, t, channels)
+			}
+		}
+	}
+}
+
+func (w *WAL) read(ctx context.Context, token string, channels *walChannels) {
+	defer w.releaseConnection() // concurrency control
+	r, err := w.walStore.Get(ctx, token)
+	defer r.Close()
+	if err != nil {
+		channels.oops <- err
+		return
+	}
+	var b []byte
+	_, err = io.ReadFull(r, b)
+	entry, err := Unmarshal(b)
+	if err != nil {
+		channels.oops <- fmt.Errorf("token: %s, err: %s", token, err)
+		return
+	}
+	channels.entry <- entry
+	return
+}
+
+func (w *WAL) collectParallelResponses(ctx context.Context, channels *walChannels) {
+	entriesMap := iradix.New().Txn()
```

Comment:
out of partially idle curiosity, what's the radix tree buying here?  space, time, both?  is it perf we need?  the complexity cost is admittedly low-ish.

Reply:
When we receive responses we need to store them sorted. It was either this or pass an ordered map of the keys to expect and then validate if we received all responses.